MF-466 - ARM docker deployment

blokovi opened the pull request:
Signed-off-by: Ivan Milošević <iva@blokovi.com>

### What does this do?
Docker deployment for arm32v7, compatible with MFX-1 gateway

### Which issue(s) does this PR fix/relate to?
Resolves #466 

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

blokovi committed:
Add arm Dockerfiles
Change version of postgres images in docker-compose to use versions with  multiarch

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
docker compose file

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
remove bin qemu file

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
remove unnecesarry comments

Signed-off-by: Ivan Milošević <iva@blokovi.com>

drasko commented on docker/Dockerfile.arm:
```diff
@@ -0,0 +1,15 @@
+FROM golang:1.10-alpine AS builder
```
This Dockerfile does not make sense, because it is identical to existing one - just ARGs are passed.

Please change Makefile to pass ARCH args that are set into existing Dockerfile. I.e. use the same Dockerfile for both architectures.

blokovi replied:
Dockerfile.arm deleted, in Makefile I made `make docker_arm32v7` that pass this args

drasko commented on mqtt/Dockerfile.arm:
```diff
@@ -0,0 +1,12 @@
+RUN npm rebuild && npm install
+
+EXPOSE 1883 8880
+
```
Please remove `qemu-arm-static` at the end, to keep the image smaller.

blokovi replied:
Removed, however new image have same size

blokovi committed:
Add make dockers_arm32v7 in Makefile

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
Remove docker/Dockerfile.arm

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
Add build for arm32v7 in ui Makefile

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
Add arm32v7 tags

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
add docker-compose for arm
remove thing-db in docker compose (use same db as user)
fix dockerfile and makefile

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
fix thing database env variable

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
Switch back to separate database for things and users

Signed-off-by: Ivan Milošević <iva@blokovi.com>

drasko commented on Makefile:
```diff
@@ -7,6 +7,7 @@ BUILD_DIR = build
 SERVICES = users things http normalizer ws coap lora influxdb-writer influxdb-reader mongodb-writer mongodb-reader cassandra-writer cassandra-reader postgres-writer postgres-reader cli bootstrap
 DOCKERS = $(addprefix docker_,$(SERVICES))
 DOCKERS_DEV = $(addprefix docker_dev_,$(SERVICES))
+DOCKERS_ARM32V7 = $(addprefix docker_arm32v7_,$(SERVICES))
```
Call it `DOCKERS_ARM` for simplicity. But this is probably not needed.

drasko commented on Makefile:
```diff
@@ -17,6 +18,10 @@ define make_docker
 	docker build --no-cache --build-arg SVC_NAME=$(subst docker_,,$(1)) --tag=mainflux/$(subst docker_,,$(1)) -f docker/Dockerfile .
 endef
 
+define make_docker_arm32v7
```
I am not sure that we need this. In my opinion everything can be controlled with env vars. You will always pass env vars in a single Dockerfile, then catch th overthere as ARGS, and then apply these ARGS as ENV vars in Dockerfile

blokovi committed:
rename arm32v7 to arm

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
remove arm32v7 tag
rename arm32v7 image names

Signed-off-by: Ivan Milošević <iva@blokovi.com>

manuio reviewed (dismissed):
LGTM

blokovi committed:
Rename command for making mqtt and ui individual images, to match other miscroservices

Signed-off-by: Ivan Milošević <iva@blokovi.com>

drasko commented on docker/Dockerfile:
```diff
@@ -1,5 +1,7 @@
 FROM golang:1.10-alpine AS builder
 ARG SVC_NAME
+ARG GOARCH
```
How is this used in Dockerfile? Looks like it is not.

blokovi replied:
It is used in `&& make $SVC_NAME \` which needs this arguments.

drasko commented on docker/Dockerfile:
```diff
@@ -1,5 +1,7 @@
 FROM golang:1.10-alpine AS builder
 ARG SVC_NAME
+ARG GOARCH
+ARG GOARM
```
Same here

blokovi replied:
It is used in `&& make $SVC_NAME \` which needs this arguments.

drasko replied:
Please set ARGs as:
1. ARCH
2. VER

blokovi replied:
This is not new variables, they are already being used and in current docs we have them like `GOARCH` and `GOARM`: https://mainflux.readthedocs.io/en/latest/dev-guide/#cross-compiling-for-arm
If I change it, I will need to change that also. do you still want this change?

blokovi committed:
Push arm docker images

Signed-off-by: Ivan Milošević <iva@blokovi.com>

blokovi committed:
fix fucntion call in ci.sh

Signed-off-by: Ivan Milošević <iva@blokovi.com>

drasko commented on Makefile:
```diff
@@ -17,6 +18,10 @@ define make_docker
 endef
 
+define make_docker_arm
+	docker build --no-cache --build-arg GOARCH=arm --build-arg GOARM=7 --build-arg SVC_NAME=$(subst docker_arm_,,$(1)) --tag=mainflux/$(subst docker_arm_,,$(1))-arm32v7 -f docker/Dockerfile .
```
ARM version is hardcoded (v7). I would advise that never the less we pass the ARGs into Docer containers via ENV variables.

BTW, I need compiling for various architectures, so please keep cross-compiling generic.

drasko commented on Makefile:
```diff
@@ -7,6 +7,7 @@ BUILD_DIR = build
 SERVICES = users things http normalizer ws coap lora influxdb-writer influxdb-reader mongodb-writer mongodb-reader cassandra-writer cassandra-reader postgres-writer postgres-reader cli bootstrap
 DOCKERS = $(addprefix docker_,$(SERVICES))
 DOCKERS_DEV = $(addprefix docker_dev_,$(SERVICES))
+DOCKERS_ARM = $(addprefix docker_arm_,$(SERVICES))
```
Not needed.

drasko commented on Makefile:
```diff
@@ -63,20 +68,32 @@ $(SERVICES):
+$(DOCKERS_DEV):
+	$(call make_docker_dev,$(@))
+
+$(DOCKERS_ARM):
```
Not needed.

drasko commented on Makefile:
```diff
@@ -63,20 +68,32 @@ $(SERVICES):
 
 dockers_dev: $(DOCKERS_DEV)
 
+dockers_arm: $(DOCKERS_ARM) docker_ui_arm docker_mqtt_arm
```
Not needed.

blokovi committed:
mergiing ui/Dockerfile

Signed-off-by: Ivan Milošević <iva@blokovi.com>

drasko reviewed (approved):
Approving so we can move on. Will be corrected in the future if needed.
